signature/cose/envelope.go

```diff
@@ -0,0 +1,587 @@
+	}
+
+	certs, ok := e.base.Headers.Unprotected[cose.HeaderLabelX5Chain].([]interface{})
+	if !ok || len(certs) == 0 {
+		return nil, &signature.InvalidSignatureError{Msg: "COSE envelope invalid certificate chain"}
+	}
+	certRaw, ok := certs[0].([]byte)
+	if !ok {
+		return nil, &signature.InvalidSignatureError{Msg: "COSE envelope malformed leaf certificate"}
+	}
+	cert, err := x509.ParseCertificate(certRaw)
+	if err != nil {
+		return nil, &signature.InvalidSignatureError{Msg: err.Error()}
+	}
+
+	// core verify process, verify integrity of COSE envelope
+	publicKeyAlg, err := getSignatureAlgorithm(cert)
+	if err != nil {
+		return nil, &signature.InvalidSignatureError{Msg: err.Error()}
+	}
+	verifier, err := cose.NewVerifier(publicKeyAlg, cert.PublicKey)
+	if err != nil {
+		return nil, &signature.InvalidSignatureError{Msg: err.Error()}
+	}
+	err = e.base.Verify(nil, verifier)
+	if err != nil {
+		return nil, &signature.SignatureIntegrityError{Err: err}
+	}
+
+	// extract content
+	return e.Content()
+}
+
+// Content implements signature.Envelope interface.
+func (e *envelope) Content() (*signature.EnvelopeContent, error) {
+	// sanity check
+	if e.base == nil {
+		return nil, &signature.InvalidSignatureError{Msg: "missing COSE signature envelope"}
+	}
+
+	payload, err := e.payload()
+	if err != nil {
+		return nil, err
+	}
+	signerInfo, err := e.signerInfo()
+	if err != nil {
+		return nil, err
+	}
+	return &signature.EnvelopeContent{
+		SignerInfo: *signerInfo,
+		Payload:    *payload,
+	}, nil
+}
+
+// Given a COSE envelope, extracts its signature.Payload.
+func (e *envelope) payload() (*signature.Payload, error) {
+	cty, ok := e.base.Headers.Protected[cose.HeaderLabelContentType]
+	if !ok {
+		return nil, &signature.InvalidSignatureError{Msg: "missing content type"}
+	}
+	var contentType string
+	if contentType, ok = cty.(string); !ok {
+		return nil, &signature.InvalidSignatureError{Msg: "content type requires tstr type"}
+	}
+	return &signature.Payload{
+		ContentType: contentType,
+		Content:     e.base.Payload,
+	}, nil
+}
+
+// Given a COSE envelope, extracts its signature.SignerInfo.
+func (e *envelope) signerInfo() (*signature.SignerInfo, error) {
+	var signerInfo signature.SignerInfo
+
+	// parse signature of COSE envelope, populate signerInfo.Signature
+	sig := e.base.Signature
+	if len(sig) == 0 {
+		return nil, &signature.InvalidSignatureError{Msg: "COSE envelope missing signature"}
+	}
+	signerInfo.Signature = sig
+
+	// parse protected headers of COSE envelope and populate related
+	// signerInfo fields
+	err := parseProtectedHeaders(e.base.Headers.Protected, &signerInfo)
+	if err != nil {
+		return nil, &signature.InvalidSignatureError{Msg: err.Error()}
+	}
+
+	// parse unprotected headers of COSE envelope
+	certs, ok := e.base.Headers.Unprotected[cose.HeaderLabelX5Chain].([]interface{})
+	if !ok || len(certs) == 0 {
+		return nil, &signature.InvalidSignatureError{Msg: "COSE envelope invalid certificate chain"}
+	}
```

Comment:
Can be removed as its being already performed by https://github.com/notaryproject/notation-core-go/blob/main/signature/internal/base/envelope.go#L188-L190

Reply:
This is an early break. The workflow is Content() -> signerInfo() -> validateEnvelopeContent() -> validateSignerInfo() -> validateCertificateChain(). The above code early breaks at the 2nd step, signerInfo().